### PR TITLE
Check validation for slashing information in a block

### LINKED ISSUE #1444: Slashing information validation
This issue is related to #1076 which is about implementing the slashing protocol. We need to split some tasks related to block header validation from issue #1076 because the tasks have many things to do and could consume some time.
 
We've been adding information about slashed validators into a block. The newly added information is as follows.

```
/// Hash of random seed of the preimages for this this height
Hash random_seed;

/// List of indices to the validator UTXO set which have not revealed the preimage
uint[] missing_validators;
```

The `random_seed` must not null in any case, which is generated from the preimages of current validators, and it must be the same as the local `random_seed` from a node itself. It's the things for the validation routine to check. In the current code of Agora, there are 2 cases the information is not valid as follows besides the case where a misbehaving validator manipulates the information.

1. In the catch-up process, there are no pre-images for generating the local random seed needed for the comparison. So we need a process to get pre-images from other nodes and add the information into the `ValidatorSet`.
2. In many unit tests or network tests, there are some cases that do not consider maintaining preimages needed for validation check and generating the `random_seed`.

As a goal of this issue, we have to add the validation routine about slashing information of the block header and make the code pass the above situations and tests.

### PR DESCRIPTION
**Pull request by linked0**
We've added information about slashed validators into a block like `random_seed` and `missing_validators`. So, when a new block is being made, we should check the validity of the slashing information. There could be a situation where preimages needed for validation are missing like in the catch-up process. So we should be able to get the preimages from the network. 

Fixes #1444

### CODE REVIEW

File: source/agora/consensus/state/ValidatorSet.d [hunk 1]
```diff
@@ -579,6 +579,48 @@ public class ValidatorSet
...
+                preimages ~= PreImageInfo(enroll_key, preimage, distance);
+            }
+        }
+        catch (Exception ex)
+        {
+            log.error("Exception occured on getPreimages: {}, heights " ~
+                "[{}..{}]", ex.msg, start_height, end_height);
+        }
+
+        return preimages;
```
Comment (hewison-chris) [inline source/agora/consensus/state/ValidatorSet.d:622]: You could consider using fold to build the array directly from the results:-
```
public PreImageInfo[] getPreimages (ulong start_height,
        ulong end_height) @trusted nothrow
    {
        PreImageInfo[] empty;
        try
        {
            auto results = this.db.execute("SELECT key, preimage, distance " ~
                "FROM validator_set WHERE enrolled_height >= ? AND " ~
                "enrolled_height <= ?",
                start_height, end_height);

            return results.fold!((preimages, row) =>
                preimages ~= PreImageInfo(
                    Hash(row.peek!(char[])(0)),
                    Hash(row.peek!(char[])(1)),
                    row.peek!ushort(2)))(empty);
        }
        catch (Exception ex)
        {
            log.error("Exception occured on getPreimages: {}, heights " ~
                "[{}..{}]", ex.msg, start_height, end_height);
        }

        return empty;
    }
```

**Review by hewison-chris — COMMENTED** [review]
You have a typo in the descritpion of the first commit. (catch-up process)

File: source/agora/consensus/state/ValidatorSet.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/state/ValidatorSet.d:622]: Good suggestion, thanks.

**Comment by linked0** [comment]
> You have a typo in the descritpion of the first commit. (catch-up process)

Fixed.

File: source/agora/test/RestoreSlashingInfo.d [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+    auto set_a = network.clients[0 .. GenesisValidators];
+    auto set_b = network.clients[GenesisValidators .. $];
+
+    // generate 18 blocks, 2 short of the enrollments expiring.
+    network.generateBlocks(Height(GenesisValidatorCycle - 2));
+    network.expectBlock(Height(GenesisValidatorCycle - 2));
```
Comment (hewison-chris) [inline source/agora/test/RestoreSlashingInfo.d]: This `expectBlock` is redundant as `generateBlocks` is already waiting for the blocks in `addBlock`. (In some tests we call expectBlock with a different range of clients)

File: source/agora/test/RestoreSlashingInfo.d [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    // Block 19 we add the freeze utxos for set_b validators
+    // prepare frozen outputs for outsider validators to enroll
+    blocks[0].spendable().drop(1)
+        .map!(txb => txb
+            .split(keys).sign(TxType.Freeze))
+        .each!(tx => set_a[0].putTransaction(tx));
+
+    network.generateBlocks(Height(GenesisValidatorCycle - 1));
+    network.expectBlock(Height(GenesisValidatorCycle - 1));
```
Comment (hewison-chris) [inline source/agora/test/RestoreSlashingInfo.d]: Same here. The clients are defaulting to the same in both calls.

File: source/agora/test/RestoreSlashingInfo.d [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    network.generateBlocks(Height(GenesisValidatorCycle));
+    network.expectBlock(Height(GenesisValidatorCycle));
+
+    // Sanity check
+    auto b20 = set_a[0].getBlocksFrom(GenesisValidatorCycle, 1)[0];
+    assert(b20.header.enrollments.length == conf.outsider_validators);
+
+    // Block 21
+    network.generateBlocks(Height(GenesisValidatorCycle + 1));
+    network.expectBlock(Height(GenesisValidatorCycle + 1));
```
Comment (hewison-chris) [inline source/agora/test/RestoreSlashingInfo.d]: Also

File: source/agora/test/RestoreSlashingInfo.d (hunk 2 above)
Comment (linked0) [reply source/agora/test/RestoreSlashingInfo.d]: I missed it. Will remove the unnecessary thing.

File: source/agora/test/RestoreSlashingInfo.d [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        retryFor(node.getBlockHeight() == GenesisValidatorCycle - 1, 2.seconds,
+            format!"Expected block height %s but outsider %s has height %s."
+                (GenesisValidatorCycle - 1, idx, node.getBlockHeight())));
+
+    // Now we enroll the set B validators.
+    set_b.enumerate.each!((idx, _) => network.enroll(GenesisValidators + idx));
+
+    // Block 20, After this the Genesis block enrolled validators will be expired.
+    network.generateBlocks(Height(GenesisValidatorCycle));
+    network.expectBlock(Height(GenesisValidatorCycle));
```
Comment (linked0) [inline source/agora/test/RestoreSlashingInfo.d]: It's interesting that removing this one makes the test fail.

Comment (hewison-chris) [reply source/agora/test/RestoreSlashingInfo.d]: Probably due to delay caused by this check. So maybe the problem is in the test code after this.

Comment (hewison-chris) [reply source/agora/test/RestoreSlashingInfo.d]: I think I see what it is. Now the default clients is no longer correct. You need to pass all the clients to the generateBlocks now there are more validators.

Comment (linked0) [reply source/agora/test/RestoreSlashingInfo.d]: Thanks for the explanation.

File: source/agora/consensus/state/ValidatorSet.d (hunk 1 above)
Comment (linked0) [reply source/agora/consensus/state/ValidatorSet.d:622]: The code you suggested outputs some error like this.

> Error: no property fold for type d2sqlite3.results.ResultRange

I have to learn more about the`fold`. It's not bad to leave the code as it is but I will try to use the `fold` in a similar context like this one. Thanks for introducing the new concept.

File: source/agora/test/RestoreSlashingInfo.d (hunk 2 above)
Comment (linked0) [reply source/agora/test/RestoreSlashingInfo.d]: Updated.

File: source/agora/test/RestoreSlashingInfo.d (hunk 3 above)
Comment (linked0) [reply source/agora/test/RestoreSlashingInfo.d]: Updated.

File: source/agora/test/RestoreSlashingInfo.d (hunk 4 above)
Comment (linked0) [reply source/agora/test/RestoreSlashingInfo.d]: Updated.

File: source/agora/test/RestoreSlashingInfo.d (hunk 5 above)
Comment (linked0) [reply source/agora/test/RestoreSlashingInfo.d]: Updated. Thanks.

File: source/agora/consensus/state/ValidatorSet.d (hunk 1 above)
Comment (hewison-chris) [reply source/agora/consensus/state/ValidatorSet.d:622]: I think you probably missed the import std.algorithm

Comment (hewison-chris) [reply source/agora/consensus/state/ValidatorSet.d:622]: Sorry I forgot to include in the sugestion.

**Comment by linked0** [comment]
Ready for review.

File: source/agora/consensus/state/ValidatorSet.d (hunk 1 above)
Comment (hewison-chris) [reply source/agora/consensus/state/ValidatorSet.d:622]: This does work:-
```
public PreImageInfo[] getPreimages (ulong start_height,
        ulong end_height) @trusted nothrow
    {
        import std.algorithm;
        PreImageInfo[] empty;
        try
        {
            auto results = this.db.execute("SELECT key, preimage, distance " ~
                "FROM validator_set WHERE enrolled_height >= ? AND " ~
                "enrolled_height <= ?",
                start_height, end_height);

            return results.fold!((preimages, row) =>
                preimages ~= PreImageInfo(
                    Hash(row.peek!(char[])(0)),
                    Hash(row.peek!(char[])(1)),
                    row.peek!ushort(2)))(empty);
        }
        catch (Exception ex)
        {
            log.error("Exception occured on getPreimages: {}, heights " ~
                "[{}..{}]", ex.msg, start_height, end_height);
        }

        return empty;
    }
```

Comment (hewison-chris) [reply source/agora/consensus/state/ValidatorSet.d:622]: No need to change now but just for reference.

Comment (linked0) [reply source/agora/consensus/state/ValidatorSet.d:622]: Right. I have already succeeded with your information yesterday. :) But, I decided to use the original code because using `foreach` is easier to understand, I think. I will apply the `fold` in a more adequate situation or context. Thanks.

File: source/agora/consensus/EnrollmentManager.d [hunk 6]
```diff
@@ -773,6 +773,25 @@ public class EnrollmentManager
...
+        Params:
+            preimage = the pre-images information to add
+
+    ***************************************************************************/
+
+    public void addPreimages (const PreImageInfo[] preimages) @safe nothrow
+    {
+        foreach (image; preimages)
+        {
+            this.validator_set.addPreimage(image);
```
Comment (AndrejMitrovic) [inline source/agora/consensus/EnrollmentManager.d]: You're ignoring the return value of `addPreimage`. What if a node sends you the wrong preimages that don't validate? Then you should exit early and try getting them from another node.

Specifically another node could send you a giant array of `PreImageInfo`. It would be good to return early if any of the preimages are invalid.

File: source/agora/consensus/SlashPolicy.d [hunk 7]
```diff
@@ -173,6 +174,12 @@ public class SlashPolicy
                 valid_keys ~= key;
         }
 
+        version (unittest)
+        {
+            if (valid_keys.length == 0)
+                return getTestRandomSeed();
```
Comment (AndrejMitrovic) [inline source/agora/consensus/SlashPolicy.d]: Why is this necessary?

File: source/agora/consensus/EnrollmentManager.d (hunk 6 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: Right, it would be better.
Will fix it.

File: source/agora/consensus/SlashPolicy.d (hunk 7 above)
Comment (linked0) [reply source/agora/consensus/SlashPolicy.d]: There are so many cases in unit tests where the pre-images needed don't exist. It's because the tests don't in fact need to focus on the existence of pre-images. So I decided to just add codes that work around errors due to missing pre-images because it will take so much time to provide proper preimages in the unit tests.

Comment (AndrejMitrovic) [reply source/agora/consensus/SlashPolicy.d]: Isn't that what [waitForPreimages](https://github.com/bpfkorea/agora/blob/c7b78c6092a03efe42598348a2cb9b10cdb9e66b/source/agora/test/Base.d#L677) and [expectBlock](https://github.com/bpfkorea/agora/blob/c7b78c6092a03efe42598348a2cb9b10cdb9e66b/source/agora/test/Base.d#L657) which calls it is for?

Comment (linked0) [reply source/agora/consensus/SlashPolicy.d]: Yes, right. But I mean unit tests, not network tests. The functions are not used in unit tests.

Comment (AndrejMitrovic) [reply source/agora/consensus/SlashPolicy.d]: I see. The issue I see is that this code will also be used by the networking tests.

Maybe a better way is to provide a custom SlashPolicy in the non-networking tests, which overrides `getRandomSeed`.

Comment (linked0) [reply source/agora/consensus/SlashPolicy.d]: I think it's better, thanks. And I agree that the code could affect the networking test.

**Review by AndrejMitrovic — APPROVED** [review]
Overall LGTM. Are the test failures just random?

**Comment by linked0** [comment]
> Overall LGTM. Are the test failures just random?

Yes, right. It randomly happens.

**Comment by linked0** [comment]
All green!

**Comment by AndrejMitrovic** [comment]
LGTM. Anyone else?

File: source/agora/consensus/EnrollmentManager.d [hunk 8]
```diff
@@ -736,6 +736,26 @@ public class EnrollmentManager
...
+            start_height = the starting enrolled height to begin retrieval from
+            end_height = the end enrolled height to finish retrieval to
+
+        Returns:
+            preimages' information of the validators
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getValidatorPreimages (ulong start_height,
+        ulong end_height) @trusted nothrow
```
Comment (Geod24) [inline source/agora/consensus/EnrollmentManager.d]: Why is it `@trusted` ?

File: source/agora/consensus/state/ValidatorSet.d [hunk 9]
```diff
@@ -579,6 +579,48 @@ public class ValidatorSet
...
+            start_height = the starting enrolled height to begin retrieval from
+            end_height = the end enrolled height to finish retrieval to
+
+        Returns:
+            preimages' information of the validators
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimages (ulong start_height,
+        ulong end_height) @trusted nothrow
```
Comment (Geod24) [inline source/agora/consensus/state/ValidatorSet.d]: Using `Height` type here

File: source/agora/network/NetworkManager.d [hunk 10]
```diff
@@ -835,8 +862,9 @@ public class NetworkManager
 
     ***************************************************************************/
 
-    private void getBlocksFrom (Height block_height,
-        scope bool delegate(const(Block)[]) @safe onReceivedBlocks) nothrow
+    private void getBlocksFrom (Ledger ledger, Height block_height,
+        scope bool delegate(Ledger ledger, const(Block)[], const(PreImageInfo)[]) @safe onReceivedBlocks)
+        nothrow
```
Comment (Geod24) [inline source/agora/network/NetworkManager.d]: I think this mixes things up quite a bit. How come we can receive a valid block but not know the pre-images ?

File: source/agora/utils/Test.d [hunk 11]
```diff
@@ -630,6 +630,12 @@ public auto genesisSpendable () @safe pure nothrow
     return GenesisBlock.spendable;
 }
 
+/// Get random seed for making new block
+public Hash getTestRandomSeed () @safe nothrow
+{
+    return hashFull(Hash.init);
+}
+
```
Comment (Geod24) [inline source/agora/utils/Test.d:665]: What is the "random seed" supposed to be ?

File: source/agora/network/NetworkManager.d (hunk 10 above)
Comment (linked0) [reply source/agora/network/NetworkManager.d]: > I think this mixes things up quite a bit. How come we can receive a valid block but not know the pre-images?

Before making a new block, we should know the pre-images as you know. So the following code will be running when the case you are saying happens which means getting blocks and pre-images from the next node. 

```
// one or more blocks were rejected, stop retrieval from node
if (!onReceivedBlocks(ledger, blocks, preimages))
    continue LNextNode;
```

I thought that there's a very low possibility that pre-images are not correct although blocks are valid.

In fact, there's one more thing we should think about it. What if there's no node from which we can get pre-images? We can imagine that there are nodes that only have the ledger and no normal database. Don't we need to store the last pre-images of un-enrolled validators into the ledger, including slashed validators? I know you don't want the ledger to be bloated due to that. I will file a new issue about that if needed.

File: source/agora/consensus/EnrollmentManager.d (hunk 8 above)
Comment (linked0) [reply source/agora/consensus/EnrollmentManager.d]: Ah, I think it should be `@safe`.

File: source/agora/consensus/state/ValidatorSet.d (hunk 9 above)
Comment (linked0) [reply source/agora/consensus/state/ValidatorSet.d]: Will update.

File: source/agora/utils/Test.d (hunk 11 above)
Comment (linked0) [reply source/agora/utils/Test.d:665]: It is supposed to be the return value of `getRandomSeed` of the `Ledger`. 

```
public Hash getRandomSeed() nothrow @safe
{
    return this.slash_man.getRandomSeed(this.last_block.header.height);
}
```

There is no pre-image that is revealed by gossiping in unit tests so the `getTestRandomSeed` is needed.

File: source/agora/network/NetworkManager.d (hunk 10 above)
Comment (linked0) [reply source/agora/network/NetworkManager.d]: > I think this mixes things up quite a bit. How come we can receive a valid block but not know the pre-images?

Anyway, we need to mix blocks and pre-images up in this function because making a new block needs proper pre-images of the current height. The fact that a block received from some node is valid means the pre-images from the node also valid because the blocks were created with the right pre-images. If there is some invalid information, the catch-up process can restart getting valid information from other nodes at the point where the failure occurs. If you have any other suggestions about that, please let me know or we can make things right with another new issue.

File: source/agora/api/FullNode.d [hunk 12]
```diff
@@ -213,6 +213,24 @@ public interface API
...
+            start_height = the starting enrolled height to begin retrieval from
+            end_height = the end enrolled height to finish retrieval to
+
+        Returns:
+            preimages' information of the validators
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimages (ulong start_height, ulong end_height);
+
```
Comment (Geod24) [inline source/agora/api/FullNode.d:233]: Why do we need this ? `getPreimage` should be enough, shouldn't it ?

File: source/agora/network/NetworkManager.d [hunk 13]
```diff
@@ -850,6 +855,19 @@ public class NetworkManager
 
             do
             {
+                auto start_height =
+                    (block_height <= 1) ?  0 : block_height;
+                auto end_height = start_height +
+                    (block_height <= 1) ?  this.consensus_config.validator_cycle :
+                        this.consensus_config.validator_cycle - 1;
```
Comment (Geod24) [inline source/agora/network/NetworkManager.d]: Why is `1` a special value here ?

File: source/agora/network/NetworkClient.d [hunk 14]
```diff
@@ -370,6 +370,29 @@ class NetworkClient
...
+        Returns:
+            the array of preimages of validators enrolling from `enrolled_height`
+            to `end_height`
+
+            If the request failed, returns an empty array
+
+    ***************************************************************************/
+
+    public PreImageInfo[] getPreimages (ulong start_height,
+        ulong end_height) @trusted nothrow
```
Comment (Geod24) [inline source/agora/network/NetworkClient.d]: Why `@trusted` ?

File: source/agora/node/FullNode.d [hunk 15]
```diff
@@ -300,16 +300,40 @@ public class FullNode : API
...
+        Params:
+            blocks = the blocks to be added
+            preimages = the preimages needed to check the validity of the blocks
+
+        Returns:
+            true if the blocks and preimages was added
+
+    ***************************************************************************/
+
+    bool addBlocks(const(Block)[] blocks, const(PreImageInfo)[] preimages) @safe
```
Comment (Geod24) [inline source/agora/node/FullNode.d]: Missing space before parenthesis

File: source/agora/node/FullNode.d [hunk 16]
```diff
@@ -300,16 +300,40 @@ public class FullNode : API
...
+    {
+        foreach (block; blocks)
+        {
+            if(!this.ledger.addPreimages(preimages))
```
Comment (Geod24) [inline source/agora/node/FullNode.d]: Instead of adding a function in the ledger whose full role is to forward to the enrollment manager, call the enrollment manager directly and remove the Ledger function

File: source/agora/consensus/SlashPolicy.d [hunk 17]
```diff
@@ -29,6 +29,7 @@ module agora.consensus.SlashPolicy;
 
 import agora.common.Amount;
 import agora.common.crypto.Key;
+import agora.common.Hash;
```
Comment (Geod24) [inline source/agora/consensus/SlashPolicy.d]: Why the extra import ?

File: source/agora/consensus/validation/Block.d [hunk 18]
```diff
@@ -179,6 +181,12 @@ public string isInvalidReason (const ref Block block, Height prev_height,
             return fail_reason;
     }
 
+    if (block.header.random_seed == Hash.init
+        || block.header.random_seed != random_seed)
+    {
+        return "Block: Random seed is wrong";
```
Comment (Geod24) [inline source/agora/consensus/validation/Block.d]: "Block: Header's random seed does not match that of known pre-images"

File: source/agora/consensus/validation/Block.d [hunk 19]
```diff
@@ -639,14 +647,17 @@ version (unittest)
...
         size_t active_enrollments, size_t enrolled_validators, scope FeeChecker checkFee,
-        scope EnrollmentFinder findEnrollment, ulong enrollment_cycle = 0,
-        ulong prev_timestamp = 0, ulong curr_timestamp = ulong.max,
+        scope EnrollmentFinder findEnrollment, Hash random_seed = Hash.init,
+        ulong enrollment_cycle = 0, ulong prev_timestamp = 0, ulong curr_timestamp = ulong.max,
         Duration block_timestamp_tolerance_dur = 100.seconds,
         string file = __FILE__, size_t line = __LINE__) nothrow @safe
     {
+        if (random_seed == Hash.init)
+            random_seed = getTestRandomSeed();
```
Comment (Geod24) [inline source/agora/consensus/validation/Block.d:657]: Why not just give it as default parameter ?

File: source/agora/utils/Test.d (hunk 11 above)
Comment (Geod24) [reply source/agora/utils/Test.d:665]: Isn't that a bug though ? That pre-images are not revealed by gossiping ?

File: source/agora/node/Ledger.d [hunk 20]
```diff
@@ -1067,6 +1071,22 @@ public class Ledger
...
+        Returns:
+            the random seed if there are one or more valid preimages,
+            otherwise Hash.init.
+
+    ***************************************************************************/
+
+    public Hash getRandomSeed() nothrow @safe
+    {
+        return this.slash_man.getRandomSeed(this.last_block.header.height);
+    }
```
Comment (Geod24) [inline source/agora/node/Ledger.d:1079]: This should be in the slashing manager directly and not in the Ledger. The problem is, we shouldn't need to do dependency injection here, because the random seed is derived from pre-images and shouldn't need to be changed. Or am I missing something ?

File: source/agora/network/NetworkManager.d (hunk 13 above)
Comment (linked0) [reply source/agora/network/NetworkManager.d]: > Why is 1 a special value here ?

I found out that `1` is not needed if I make the formula more correct and simpler.
I changed it. Ty.

File: source/agora/consensus/validation/Block.d (hunk 19 above)
Comment (linked0) [reply source/agora/consensus/validation/Block.d:657]: > Why not just give it as default parameter ?

It's because there are some cases the `random_seed` is already set before calling the `isValidcheck` function. In many test cases, there are two situations where the `random_seed` is set or not.

File: source/agora/utils/Test.d (hunk 11 above)
Comment (linked0) [reply source/agora/utils/Test.d:665]: > Isn't that a bug though ? That pre-images are not revealed by gossiping ?

No, it's not. It's only used in unit tests without any gossiping.

File: source/agora/node/Ledger.d (hunk 20 above)
Comment (linked0) [reply source/agora/node/Ledger.d:1079]: I understand what you mean. But Getting the random seed is needed in unit tests like in `TestLedger` so I've got to make this function overridden which is used as follows.

```
static class ThrowingLedger : Ledger
{
     ///
     public override Hash getRandomSeed() nothrow @safe
     {
         return getTestRandomSeed();
     }
}
```

**Comment by linked0** [comment]
Applied all the suggestions and rebased. Waiting for all green. 
Thanks for the review.

File: source/agora/consensus/validation/Block.d (hunk 19 above)
Comment (Geod24) [reply source/agora/consensus/validation/Block.d:657]: But if it's set, it can be passed as a parameter ? Maybe you have a link to the code that creates the issue ?

Comment (linked0) [reply source/agora/consensus/validation/Block.d:657]: > But if it's set, it can be passed as a parameter? Maybe you have a link to the code that creates the issue?

I'll make the value be passed as a parameter. I think it's the last thing to do in this PR.

**Review by Geod24 — APPROVED** [review]
